backend-tests/tests/test_artifact.py

```diff
@@ -0,0 +1,352 @@
+            ) as artifact:
+                r = api_client.with_auth(auth_token).call(
+                    "POST",
+                    deployments.URL_DEPLOYMENTS_ARTIFACTS,
+                    files=(
+                        ("description", (None, "description")),
+                        ("size", (None, str(os.path.getsize(artifact)))),
+                        (
+                            "artifact",
+                            (
+                                artifact,
+                                open(artifact, "rb"),
+                                "application/octet-stream",
+                            ),
+                        ),
+                    ),
+                )
+            assert r.status_code == 201
+
+        # create a new accepted device
+        devauthd = ApiClient(deviceauth_v1.URL_DEVICES)
+        dev = make_accepted_device(auth_token, devauthd, tenant.tenant_token)
+        assert dev is not None
+
+        # create a deployment
+        resp = api_client.with_auth(auth_token).call(
+            "POST",
+            deployments.URL_DEPLOYMENTS,
+            body={
+                "name": "deployment-1",
+                "artifact_name": "test",
+                "devices": [dev.id],
+            },
+        )
+        assert resp.status_code == 201
+
+        return dev
+
+    def test_upload_artifact_selection_no_match(self, mongo, clean_mongo):
+        dev = self.setup_upload_artifact_selection()
+        deploymentsd = ApiClient(deployments.URL_DEVICES)
+        r = deploymentsd.with_auth(dev.token).call(
+            "POST",
+            deployments.URL_NEXT,
+            body={"device_type": "arm1", "artifact_name": "old-artifact"},
+        )
+        assert r.status_code == 204
+
+    def test_upload_artifact_selection_no_match_wrong_depends(self, mongo, clean_mongo):
+        dev = self.setup_upload_artifact_selection(
+            checksums=("checksum", "another-checksum",)
+        )
+        deploymentsd = ApiClient(deployments.URL_DEVICES)
+        r = deploymentsd.with_auth(dev.token).call(
+            "POST",
+            deployments.URL_NEXT,
+            body={
+                "device_type": "arm1",
+                "artifact_name": "old-artifact",
+                "rootfs_image_checksum": "wrong-checksum",
+            },
+        )
+        assert r.status_code == 204
+
+    def test_upload_artifact_selection_match_depends(self, mongo, clean_mongo):
+        dev = self.setup_upload_artifact_selection()
+        deploymentsd = ApiClient(deployments.URL_DEVICES)
+        r = deploymentsd.with_auth(dev.token).call(
+            "POST",
+            deployments.URL_NEXT,
+            body={
+                "device_type": "arm1",
+                "artifact_name": "old-artifact",
+                "rootfs_image_checksum": "checksum",
+            },
+        )
+        assert r.status_code == 200
+
+    def test_upload_artifact_selection_match_depends_multiple_artifacts(
```

Comment:
It would be useful to have a testcase where the provides satisfies multiple artifacts and verify that the artifact with the smallest file-size is picked. For instance, make a deployment with artifact depends:
```
{
    "device_type": ["acme"],
    "vcdiff": "xdelta"
} (artifact_name: "common", size 100KiB)
```
```
{
    "device_type": ["acme"]
} (artifact_name: "common", size: 256MiB)
```
And a POST /next request with provides
```
{
    "artifact_name": "foobar",
    "device_type": "acme",
    "vcdiff": "xdelta"
}
```
And check the first is picked.

Reply:
Done!